src/transforms/kubernetes/mod.rs

```diff
@@ -288,124 +328,89 @@ impl Transform for KubernetesPodMetadata {
-    let fun = move |pod: &Pod| {
-        fun(pod)
-            .map(|data| vec![(key.clone(), data.into())])
-            .unwrap_or_default()
-    };
-    (name, Box::new(fun) as Box<_>)
-}
+    /// Extracts this field from Pod
+    fn extract(self, pod: &Pod) -> Option<Vec<(Atom, Value)>> {
+        use Field::*;
+
+        Some(match self {
+            // ------------------------ ObjectMeta ------------------------ //
+            Name => self.field(pod.metadata.as_ref()?.name.clone()?),
+            Namespace => self.field(pod.metadata.as_ref()?.namespace.clone()?),
+            CreationTimestamp => self.field(pod.metadata.as_ref()?.creation_timestamp.clone()?.0),
+            DeletionTimestamp => self.field(pod.metadata.as_ref()?.deletion_timestamp.clone()?.0),
+            Labels => self.collection(pod.metadata.as_ref()?.labels.as_ref()?),
+            Annotations => self.collection(pod.metadata.as_ref()?.annotations.as_ref()?),
+            // ------------------------ PodSpec ------------------------ //
+            NodeName => self.field(pod.spec.as_ref()?.node_name.clone()?),
+            Hostname => self.field(pod.spec.as_ref()?.hostname.clone()?),
+            Priority => self.field(pod.spec.as_ref()?.priority.clone()?),
+            PriorityClassName => self.field(pod.spec.as_ref()?.priority_class_name.clone()?),
+            ServiceAccountName => self.field(pod.spec.as_ref()?.service_account_name.clone()?),
+            Subdomain => self.field(pod.spec.as_ref()?.subdomain.clone()?),
+            // ------------------------ PodStatus ------------------------ //
+            HostIp => self.field(pod.status.as_ref()?.host_ip.clone()?),
+            Ip => self.field(pod.status.as_ref()?.pod_ip.clone()?),
+        })
+    }
 
-fn collection_field(
-    name: &'static str,
-    fun: impl Fn(&Pod) -> Option<&BTreeMap<String, String>> + Send + Sync + 'static,
-) -> (
-    &'static str,
-    Box<dyn Fn(&Pod) -> Vec<(Atom, Value)> + Send + Sync + 'static>,
-) {
-    let prefix_key = with_prefix(name) + ".";
-    let fun = move |pod: &Pod| {
-        fun(pod)
-            .map(|map| {
-                map.iter()
-                    .map(|(key, value)| ((prefix_key.clone() + key).into(), value.into()))
-                    .collect()
-            })
-            .unwrap_or_default()
-    };
-    (name, Box::new(fun) as Box<_>)
-}
+    fn field(self, data: impl Into<Value>) -> Vec<(Atom, Value)> {
+        vec![(Self::with_prefix(self.name()).into(), data.into())]
+    }
+
+    fn collection(self, map: &BTreeMap<String, String>) -> Vec<(Atom, Value)> {
+        let prefix_key = Self::with_prefix(self.name()) + ".";
+
+        map.iter()
+            .map(|(key, value)| ((prefix_key.clone() + key).into(), value.into()))
+            .collect()
+    }
 
-fn with_prefix(name: &str) -> String {
-    event::log_schema().kubernetes_key().as_ref().to_owned() + "." + name
+    fn with_prefix(name: String) -> String {
```

Comment:
technically this could be a `&str` since we deref into `&name` below

Reply:
True